ports/rp2/main.c

```diff
@@ -61,6 +68,50 @@ bi_decl(bi_program_feature_group_with_flags(BINARY_INFO_TAG_MICROPYTHON,
     BINARY_INFO_ID_MP_FROZEN, "frozen modules",
     BI_NAMED_GROUP_SEPARATE_COMMAS | BI_NAMED_GROUP_SORT_ALPHA));
 
+#if MICROPY_VFS_FAT && MICROPY_HW_USB_MSC
+#define debug_printf(...) mp_printf(&mp_plat_print, "main.c: " __VA_ARGS__)
+
+int tud_msc_device_ejected_cb(bool ejected) {
+    if (mp_initialized == false) {
+        // Make sure this is not called before MP is initialized.
+        return 0;
+    }
+
+    mp_uint_t atomic_state = MICROPY_BEGIN_ATOMIC_SECTION();
+    int ret = 0;
+    const char *mount_point_str = "/flash";
+    mp_obj_t mount_point = mp_obj_new_str(mount_point_str, strlen(mount_point_str));
+    nlr_buf_t nlr;
+    if (nlr_push(&nlr) == 0) {
+        if (ejected) {
+            // Mount filesystem to allow device access.
+            mp_obj_t bdev = rp2_flash_type.make_new(&rp2_flash_type, 0, 0, NULL);
+            mp_obj_t args[] = { bdev, mount_point, MP_OBJ_NEW_QSTR(MP_QSTR_mkfs), mp_const_true };
+            mp_map_t kw_args;
+            mp_map_init_fixed_table(&kw_args, 1, args + 2);
+            mp_vfs_mount(2, args, &kw_args);
+            mp_vfs_chdir(mount_point);
+        } else {
+            // Unmount filesystem to allow host access.
+            mp_vfs_umount(mount_point);
```

Comment:
I'm not sure how well this will work if code is currently executing from this mount point, and files are open.

Instead of mounting/unmounting dependent on the ejected state, maybe just make the FS read-only when not ejected?

Reply:
This gets called from main too before executing any code, and mounts the filesystem if the board is Not connected and was Not ejected, otherwise it mounts the filesystem for the device see line 172:
```
tud_msc_device_ejected_cb(!tud_connected() || tud_msc_device_ejected());
```
> maybe just make the FS read-only when not ejected?

I'm not sure it's possible to make it read-only for the device side and read/write for host, but the device filesystem will be outdated, i.e. host can be make changes to the fs, remove files etc.

Note I should use the protected mount instead, and if possible there should be a matching protected umount.

Reply:
Note you can drop the last commit if not needed, for the old (stm32-compatible) behavior.

Reply:
Isn't this function called from the TinyUSB callback, if the MSC is ejected?  And so it can be executed at any time (asynchronously), eg while code is running?

Reply:
Yes, it can be called at any time by TinyUSB when MSC is ejected. It will then mount the filesystem in an atomic section. Note it's called in main *synchronously* to re/mount or unmount the filesystem (if USB is Not connected to a host, or MSC has been ejected by host) on first/and subsequent soft-boots before `main.oy`, `boot.py`, REPL or any filesystem script is executed.
```
// Mount flash filesystem for device if not connected to a host.
tud_msc_device_ejected_cb(!tud_connected() || tud_msc_device_ejected());
```

Reply:
I kept the last commit that adds device/host mutual exclusion for reference, maybe we need it later, but feel free to drop it.